src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,266 @@
+    auto state = co_await invoke_on_all_nodes(
+      [req](auto& handle) { return handle->start_test(req); });
+    co_return id;
+}
+
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+    int8_t retries = 3;
+    auto test_state = co_await status();
+    while (!test_state.finished() && retries-- > 0) {
+        /// invoke_on_all will filter brokers that have shut down cleanly
+        /// from its participants list, retries (if occur) will only occur
+        /// on previously unreachable nodes
+        test_state = co_await invoke_on_all_nodes(
+          [](auto& handle) { return handle->stop_test(); });
+        if (!test_state.finished()) {
+            vlog(
+              clusterlog.warn,
+              "Failed to sucessfully stop all outstanding jobs on all "
+              "participants: {} , trying again, retries: {}",
+              test_state,
+              retries);
+        }
+        co_await ss::sleep(1s);
+    }
+    co_return test_state;
+}
+
+ss::future<self_test_frontend::global_test_state> self_test_frontend::status() {
+    return invoke_on_all_nodes(
+      [](auto& handle) { return handle->get_status(); });
+}
+
+template<typename Func>
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::invoke_on_all_nodes(Func f) {
+    auto gate_holder = _gate.hold();
+    auto brokers = [this]() {
+        /// Goal is to return either
+        /// 1. All brokers if this is a new test run
+        /// 2. Only ids of brokers who haven't yet returned results if test
+        ///    has not ended
+        auto participants = _pending_run.active_participant_ids();
+        return participants.empty() ? _members.local().node_ids()
+                                    : participants;
+    }();
+    auto local_state_vec = co_await ssx::parallel_transform(
+      brokers, [this, f = std::move(f)](const auto& node_id) {
+          std::unique_ptr<invoke_wrapper> handle;
+          if (_self == node_id) {
+              handle = std::make_unique<local_invoke>(node_id, _self_be);
+          } else {
+              handle = std::make_unique<remote_invoke>(
+                node_id, _self, _connections);
+          }
+          return ss::do_with(
+                   std::move(handle),
+                   [f = std::move(f)](auto& handle) { return f(handle); })
+            .then([node_id](node_test_state result) {
+                return std::pair<const model::node_id, node_test_state>(
+                  node_id, std::move(result));
+            });
+      });
+    auto result = global_test_state(std::move(local_state_vec));
+
+    /// Condense response from all nodes to new known state of the system
+    _pending_run.merge(std::move(result));
+
+    if (!_pending_run.finished()) {
+        co_return _pending_run;
+    }
+    auto previous_successful_run = _pending_run;
+    _pending_run = global_test_state({});
+    co_return previous_successful_run;
```

Comment:
I'm liking how we're getting to be more stateless now that we're no longer keeping the previous run around. Could we take this even further by not tracking `_pending_run` at all, and instead always use the full set of node IDs returned by the members table?

As implemented, we're saving on some duplicate RPCs, but we also discussed that the RPCs were pretty lightweight given the caching within the test workers. I think it'd be easier to reason about without knowing the nitty-gritty of this implementation, though maybe I'm missing something

Reply:
Recently removed as per offline discussion